### DISCUSSION TITLE
[@mantine/form] Allow usage of asynchronous validation with zod resolver

### DISCUSSION THREAD

**Discussion by Punie** [mantinedev/mantine#2078] [Feature requests]
### What package has an issue

@mantine/form

### Describe the bug

First of all, thank you for the amazing work done on Mantine 4.0 🙏 

I my current project, I am making heavy use of form validation through `zod` and in some cases I have a need for asynchronous validation, eg:
```ts
const signupSchema = z
  .object({
    username: z
      .string()
      .refine(
        value =>
          axios
            .post('/api/check_unique_username', { username: value })
            .then(() => true)
            .catch(() => false),
        {
          message: 'Username already exists',
        },
      ),
  });
```

However, this doesn't work with the current implementation of `zodResolver` because it uses the `safeParse` method (see [here](https://github.com/mantinedev/mantine/blob/master/src/mantine-form/src/resolvers/zod-resolver/zod-resolver.ts#L23)). 

### In which browser did the problem occur

All of them

### Do you know how to fix the issue

Yes

### Are you willing to participate in fixing this issue and create a pull request with the fix

Yes

### Possible fix

Replace the call to `safeParse` by a call to [`safeParseAsync`](https://github.com/colinhacks/zod#safeparseasync) in `zodResolver`

**Comment by rtivital** [comment]
You can send a PR to the `master` branch

**Comment by maticrivo** [comment]
Is anyone working on this? i see https://github.com/mantinedev/mantine/pull/1255 is closed..

**Comment by HananoshikaYomaru** [comment]
I also think handling async validation should be support.

**Comment by zogot** [comment]
I have solved the requirement for async validation for a username check by doing the following:

Can probably have better types, but figured I'll show what I've been doing for others who may also be looking.

```typescript
    const checkAvailability = async (event: FocusEvent) => {
        const result = await fetch('/api/user/check-username', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            // @ts-ignore
            body: JSON.stringify({ username: event.target.value })
        })

        const resultBody = await result.json();
        if(resultBody.available) {
            return true;
        }

        form.setFieldError('email', "Username taken")
    }

    const { onBlur: formOnBlur, ...emailProps} = form.getInputProps('email')
    const onBlur = async (event) => {
        await formOnBlur(event)
        await checkAvailability(event)
    }
```
```typescript
<TextInput 
    size='md' 
    mt='md'
    withAsterisk 
    label={LL.signup.fields.email.label()} 
    placeholder={LL.signup.fields.email.placeholder()} 
    { ...emailProps } 
    onBlur={onBlur}
/>
```

**Reply by filipjnc** [reply]
I follow a similar approach, however it seems like `form.isValid()` is still true after `form.setFieldError()`. I describe this here: https://github.com/mantinedev/mantine/discussions/3029

There is also no helper like `form.hasErrors()` that I can use together with `form.isValid()`, e.g. to disable the form submit button.

**Reply by jstpierref** [reply]
I have the exact same use case. Ended adding `Object.keys(form.errors).length !== 0` along `isDirty` and `isValid` to disable the button.

**Reply by toofff** [reply]
From what I see in the code, the `isValid` method will restart the validation of the form via the rules set up in the `validation` attribute of the `useForm` hook (which I find normal). So if we add an error like what is put above, the `isValid` method will never be aware of it.

Even if @jstpierref's solution works, for me it is only temporary and the `zodResolver` should really be able to work asynchronously.

**Comment by yanmarinich** [comment]
I'm also looking forward to have this feature implemented as a workaround for a time being:

I made a function like this

```javascript
export const validateAsync = async <T>(
  schema: z.ZodType<T>,
  values: T,
  form: UseFormReturnType<any extends T ? T : any>,
) => {
  const result = await schema.safeParseAsync(values);

  if (!result.success) {
    form.setErrors({ ...result.error.formErrors.fieldErrors });
  }

  return result;
};
```

I created another async schema for fields that need to be validated async

```javascript
export const collectionAsyncSchema = z.object({
  symbol: z.string().refine(
    async (symbol) => {
      // verify that symbol exists in database
      return true;
    },
    {
      message: "Symbol already exists",
    },
  ),
  priceEth: z.number().refine(
      async (price) => {
        // verify price async
        return false;
      },
      {
        message: "Price is too high",
      },
  ),
});
```

and my onSubmit function looks like this:

```javascript
  const onSubmit = form.onSubmit(async (values) => {
    const result = await validateAsync(collectionAsyncSchema, values, form);

    console.log(result.success);
  });
```
